### PR TITLE
add video [Model] Video Frame Sparsification via Pixel-Level Similarity for Efficient Multimodal Long-Video Inference for Qwen2.5_vl and Qwen3_vlsparse

### PR DESCRIPTION
**Pull request by rayleeku**
## Purpose
RFC: [#31803](https://github.com/vllm-project/vllm/issues/31803)

The exponential growth of video content and the rising demand for long-video intelligence (e.g., video summarization, multimodal QA, video generation) have made long-video inference a standard requirement for multimodal large models. Current models suffer from critical efficiency bottlenecks and face two key challenges: 

temporal and spatial redundancy of videos and input sequence overflow, thus necessitating a lightweight, adaptive, and scalable sparsification solution. 

This PR proposes a video sparsification scheme that adaptively selects key frames based on video inter-frame similarity and user-defined sparsity rates, and it has been compatible with Qwen2.5-VL and Qwen3-VL.

## Implementation details

This solution implements adaptive sparsification through three stages:
1.Downsampling Preprocessing: Apply spatial downsampling (e.g., 4× reduction) to raw frames to lower computational load.
2.Similarity Assessment: Compute SSIM + L1 loss between adjacent frames to generate inter-frame similarity sequences.
3.Adaptive Keyframe Extraction:
Calculate retained frame count: K = int(total_frames × sparsity_ratio) based on user-defined sparsity ratio
Apply constraint: K = max(2, min(total_frames, 2 × (K // 2))) → ensures K is even and ≥2
Select Top-(K-1) lowest-similarity frames as segmentation points to divide the video into K segments
Extract the temporal midpoint frame from each segment as the keyframe; discard others.

## Test Plan
- [ ] Add the variable video_sparse_rate when launching the service.
- [ ] Test the performance and functionality of Qwen2.5-VL/Qwen3-VL separately when video input is provided， online and offline.

### CODE REVIEW

**Comment by rayleeku** [comment]
@DarkLight1337 I reopen a PR for this feature

File: vllm/multimodal/video_sparse.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        """
+        Calculate photometric loss (weighted SSIM + L1 loss) between two frames.
+        Lower loss means more similar frames.
+        """
+        ssim = self._calculate_ssim(frame1, frame2)
+        # Compute L1 loss (pixel-wise absolute difference)
+        if frame1.shape[0] == 3:
+            l1_loss = torch.mean(torch.abs(frame1 - frame2).float()).item()
+        else:
+            l1_loss = torch.mean(torch.abs(frame1[0] - frame2[0])).item()
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: Can use `torch.nn.functional.l1_loss` directly

Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: And there is no need to differentiate between grayscale and RGB, since you're taking the mean anyway

File: vllm/multimodal/video_sparse.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+    def _calculate_ssim(self, img1: torch.Tensor, img2: torch.Tensor) -> float:
+        """
+        Calculate Structural Similarity Index (SSIM) between two frames.
+        Higher SSIM means more similar frames (range: 0-1).
+        """
+        assert img1.shape == img2.shape
+        assert img1.device == img2.device
+
+        if img1.shape[0] == 3:
+            gray1 = 0.299 * img1[0] + 0.587 * img1[1] + 0.114 * img1[2]
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: Can use torchvision to convert from rgb into grayscale

File: vllm/multimodal/video_sparse.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        return loss
+
+    def _calculate_video_photometric_losses(self, frames: torch.Tensor) -> torch.Tensor:
+        """Compute photometric loss for all adjacent frame pairs in video."""
+        frame_number = frames.shape[0]
+        losses = []
+        for f in range(frame_number - 1):
+            loss = self._calculate_photometric_loss(frames[f], frames[f + 1])
+            losses.append(loss)
+        return torch.tensor(losses, device=frames.device)
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]:
```suggestion
        return torch.stack(losses)
```

File: vllm/inputs/preprocess.py [hunk 4]
```diff
@@ -688,6 +688,29 @@ def preprocess(
         *,
         mm_uuids: MultiModalUUIDDict | None = None,
     ) -> ProcessorInputs:
+        """Efficient Frame Selection For Videos."""
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py]: IMO this should be applied inside or after `BaseProcessingInfo.parse_mm_data`.

File: vllm/multimodal/video_sparse.py (hunk 3 above)
Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: But tbh, I suggest batching the loss computation instead of looping over them, like

```python
        ssim = self._calculate_ssim_batch(frames[:-1], frames[1:])
        l1_loss = F.l1_loss(frames[:-1], frames[1:])

        return self.alpha * (1 - ssim) / 2 + (1 - self.alpha) * l1_loss
```

Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: To optimize further:

```py
frames_grayscale = (
    frames if frames.shape[1] == 1 else tvF.rgb_to_grayscale(frames)
)
ssim = structural_similarity_index_measure(
    frames_grayscale[:-1],
    frames_grayscale[1:],
    sigma=1.5,
    kernel_size=11,
    data_range=255.0,
) 

l1_loss = F.l1_loss(frames[:-1], frames[1:])

return self.alpha * (1 - ssim) / 2 + (1 - self.alpha) * l1_loss
```

File: vllm/multimodal/video_sparse.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+import numpy as np
+import torch
+import torchvision.transforms.v2.functional as tvF
+from torchvision.ops import structural_similarity_index_measure as ssim
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: Btw I think this should be from `torchmetrics.functional.image`? I can't find this function inside `torchvision.ops`

Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: thanks I have corrected it

File: vllm/multimodal/video_sparse.py (hunk 3 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: I have modified the code accordingly. Thanks

File: vllm/multimodal/video_sparse.py (hunk 2 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: I have changed to use torchvision for the change. thanks

File: vllm/multimodal/video_sparse.py (hunk 1 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: modified accordingly

File: vllm/multimodal/video_sparse.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        return downsampled
+
+    def _calculate_ssim(self, imgs1: torch.Tensor, imgs2: torch.Tensor) -> torch.Tensor:
+        """Calculate SSIM for batch of frames."""
+        assert imgs1.shape == imgs2.shape
+        assert imgs1.device == imgs2.device
+        assert imgs1.dim() == 4, "Input must be 4D tensor (B, C, H, W)"
+
+        gray1 = tvF.rgb_to_grayscale(imgs1, num_output_channels=1).to(torch.float32)
+        gray2 = tvF.rgb_to_grayscale(imgs2, num_output_channels=1).to(torch.float32)
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: There is a lot of overlap between the two batches, so I suggest converting all the original frames to grayscale before extracting slices (i.e. prev frame, next frame) from the tensor

Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: Also why do we need to convert to fp32? Don't the two batches already have the same dtype since they originate from the same tensor?

File: vllm/multimodal/video_sparse.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        prev_frames = frames[:-1]
+        next_frames = frames[1:]
+
+        return self._calculate_photometric_loss(prev_frames, next_frames)
+
+    def _select_split_points(
+        self, photometric_losses: torch.Tensor, k: int
+    ) -> list[int]:
+        """Select top-k largest loss indices as split points."""
+        if k - 1 <= 0:
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]:
```suggestion
        if k <= 1:
```

File: vllm/multimodal/video_sparse.py (hunk 6 above)
Comment (rayleeku) [reply vllm/multimodal/video_sparse.py]: Done. The input value of the structural_similarity_index_measure function is float. When the qwen2.5-vl and 3-vl services are started online, the input value format is torch.uint8.

Comment (DarkLight1337) [reply vllm/multimodal/video_sparse.py]: I think simply casting the values is incorrect as torchvision expects float dtype to have range in `[0, 1]`, while uint8 has values in range `[0, 255]`. You should use `tvF.convert_image_dtype` to convert the image dtype.

File: vllm/inputs/preprocess.py [hunk 8]
```diff
@@ -555,6 +556,30 @@ def preprocess(
...
+        if (
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py]: Let's apply this inside `_process_multimodal` after `mm_items = mm_processor.info.parse_mm_data(mm_data)`

Comment (rayleeku) [reply vllm/inputs/preprocess.py]: modified accordingly

File: vllm/inputs/preprocess.py [hunk 9]
```diff
@@ -209,6 +210,26 @@ def _process_multimodal(
         if mm_processor_kwargs is None:
             mm_processor_kwargs = {}
 
+        # Efficient Frame Selection For Videos.
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py]: Call this on `mm_items`, not `mm_data`

Comment (rayleeku) [reply vllm/inputs/preprocess.py]: i think mm_data is better, cause mm_items need to update the metadata as well.

Comment (rayleeku) [reply vllm/inputs/preprocess.py]: may i know the reason to choose mm_items instead of mm_data, thanks

Comment (DarkLight1337) [reply vllm/inputs/preprocess.py]: We should avoid having duplicate code to parse the videos

Comment (DarkLight1337) [reply vllm/inputs/preprocess.py]: The length of `metadata_lst` is assumed to have the same length as videos, so it should be straightforward to update it by selecting the corresponding frame indices.

Comment (rayleeku) [reply vllm/inputs/preprocess.py]: I have modified accordingly. thanks

File: vllm/multimodal/video_sparse.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+                updated_metadata = video_metadata.copy()
+                updated_metadata["frames_indices"] = selected_indices
+                result.append((video_sampled[i], updated_metadata))
+            return result
+        else:
+            return video_sampled
+
+
+def is_multimodal_efs_enabled(efs_sparse_rate: float | None) -> bool:
```
Comment (DarkLight1337) [inline vllm/multimodal/video_sparse.py]: I think this method is rather redundant, you can inline this in `inputs/preprocess.py`

File: vllm/inputs/preprocess.py [hunk 11]
```diff
@@ -210,6 +211,36 @@ def _process_multimodal(
...
+            and self.model_config.multimodal_config
+        ):
+            efs_sparse_rate = self.model_config.multimodal_config.video_sparse_rate
+        else:
+            efs_sparse_rate = 0.0
+        efs_sparse_enabled = is_multimodal_efs_enabled(efs_sparse_rate)
+        if (
+            efs_sparse_enabled and "video" in mm_items  # type: ignore[typeddict-item]
+        ):
+            videos = mm_items["video"].data
```
Comment (DarkLight1337) [inline vllm/inputs/preprocess.py]:
```suggestion
            video_data = mm_items["video"]
            if isinstance(video_data, VideoProcessorItems):
                ...
```

You can add methods to `VideoProcessorItems` to access and update the video data, instead of modifying it externally.
